### PR TITLE
[y_cable] Support for initialization of new daemon ycable to support ycables

### PR DESCRIPTION
**Pull request by vdahiya12**
Signed-off-by: vaibhav-dahiya <vdahiya@microsoft.com>

This PR also adds the commit in sonic-platform-daemons

94fa239  [y_cable] refactor y_cable to a seperate logic and new daemon from xcvrd (#219)

#### Why I did it

This PR separates the logic of Y-Cable from xcvrd. Before this change we were utilizing xcvrd daemon to control all aspects of Y-Cable right from initialization to processing requests from other entities like orch,linkmgr. 
Now we would have another daemon ycabled which will serve this purpose.
Logically everything still remains the same from the perspective of other daemons. 
it also take care aspects like init/delete daemon from Y-Cable perspective. 

#### How I did it

To serve the purpose we build a new wheel sonic_ycabled-1.0-py3-none-any.whl and install it inside pmon.
We also initalize the daemon ycabled which serves our purpose for refactor inside pmon

#### How to verify it

Ran the changes with an image for dualtor tests on a 7050cx3 platform

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

### CODE REVIEW

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 [hunk 1]
```diff
@@ -108,6 +108,25 @@ dependent_startup=true
 dependent_startup_wait_for=rsyslogd:running
 {% endif %}
 
+{% if 'subtype' in DEVICE_METADATA['localhost'] and DEVICE_METADATA['localhost']['subtype'] == 'DualToR' %}
+{% if not skip_xcvrd %}
```
Comment (prgeor) [inline dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2:112]: why this check 'skip_xcvrd'? how is this new daemon dependent on xcvrd?

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 [hunk 2]
```diff
@@ -108,6 +108,25 @@ dependent_startup=true
...
+[program:ycable]
+{% if delay_xcvrd %}
```
Comment (prgeor) [inline dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2:114]: why delay start of ycable daemon based upon 'delay_xcvrd'?

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 [hunk 3]
```diff
@@ -108,6 +108,25 @@ dependent_startup=true
...
+command=bash -c "sleep 30 && {% if API_VERSION == 3 and 'ycable' not in python2_daemons %}nice -n -20 python3 {% else %} nice -n -20 python2 {% endif %}/usr/local/bin/ycable"
+{% else %}
+command=nice -n -20 {% if API_VERSION == 3 and 'ycable' not in python2_daemons %}python3 {% else %} python2 {% endif %}/usr/local/bin/ycable
+{% endif %}
+priority=6
```
Comment (prgeor) [inline dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2:119]: you seem to be reusing the 'priority' level of xcvrd. is this intentional?

File: dockers/docker-platform-monitor/docker_init.j2 [hunk 4]
```diff
@@ -105,9 +105,9 @@ confvar="{\"HAVE_SENSORS_CONF\":$HAVE_SENSORS_CONF, \"HAVE_FANCONTROL_CONF\":$HA
 
 if [ -e $PMON_DAEMON_CONTROL_FILE ];
 then
-    sonic-cfggen -j $PMON_DAEMON_CONTROL_FILE -a "$confvar" -t $SUPERVISOR_CONF_TEMPLATE > $SUPERVISOR_CONF_FILE
+    sonic-cfggen -d -j $PMON_DAEMON_CONTROL_FILE -a "$confvar" -t $SUPERVISOR_CONF_TEMPLATE > $SUPERVISOR_CONF_FILE
```
Comment (prgeor) [inline dockers/docker-platform-monitor/docker_init.j2:108]: why did you add the new option '-d' ?

File: rules/sonic-ycable.mk [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# sonic-ycable (SONiC Y-Cable daemon) Debian package
+
+# SONIC_YCABLE_PY2 package
+
+SONIC_YCABLE_PY2 = sonic_ycable-1.0-py2-none-any.whl
+$(SONIC_YCABLE_PY2)_SRC_PATH = $(SRC_PATH)/sonic-platform-daemons/sonic-ycable
+$(SONIC_YCABLE_PY2)_DEPENDS = $(SONIC_PY_COMMON_PY2) $(SONIC_PLATFORM_COMMON_PY2)
+$(SONIC_YCABLE_PY2)_DEBS_DEPENDS = $(LIBSWSSCOMMON) $(PYTHON_SWSSCOMMON)
+$(SONIC_YCABLE_PY2)_PYTHON_VERSION = 2
+SONIC_PYTHON_WHEELS += $(SONIC_YCABLE_PY2)
```
Comment (prgeor) [inline rules/sonic-ycable.mk:10]: We are moving to PY3 only, then why PY2 support?

File: rules/sonic-ycable.dep [hunk 6]
```diff
@@ -0,0 +1,16 @@
+SPATH       := $($(SONIC_YCABLE_PY2)_SRC_PATH)
+DEP_FILES   := $(SONIC_COMMON_FILES_LIST) rules/sonic-ycable.mk rules/sonic-xcvrd.dep
+DEP_FILES   += $(SONIC_COMMON_BASE_FILES_LIST)
+SMDEP_FILES := $(addprefix $(SPATH)/,$(shell cd $(SPATH) && git ls-files))
+
+$(SONIC_YCABLE_PY2)_CACHE_MODE  := GIT_CONTENT_SHA
+$(SONIC_YCABLE_PY2)_DEP_FLAGS   := $(SONIC_COMMON_FLAGS_LIST)
+$(SONIC_YCABLE_PY2)_DEP_FILES   := $(DEP_FILES)
+$(SONIC_YCABLE_PY2)_SMDEP_FILES := $(SMDEP_FILES)
+$(SONIC_YCABLE_PY2)_SMDEP_PATHS := $(SPATH)
```
Comment (prgeor) [inline rules/sonic-ycable.dep:10]: We are moving to PY3 only, then why PY2 support?

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 [hunk 7]
```diff
@@ -108,6 +108,25 @@ dependent_startup=true
 dependent_startup_wait_for=rsyslogd:running
 {% endif %}
 
+{% if 'subtype' in DEVICE_METADATA['localhost'] and DEVICE_METADATA['localhost']['subtype'] == 'DualToR' %}
+{% if not skip_xcvrd %}
+[program:ycable]
+{% if delay_xcvrd %}
+command=bash -c "sleep 30 && {% if API_VERSION == 3 and 'ycable' not in python2_daemons %}nice -n -20 python3 {% else %} nice -n -20 python2 {% endif %}/usr/local/bin/ycable"
+{% else %}
+command=nice -n -20 {% if API_VERSION == 3 and 'ycable' not in python2_daemons %}python3 {% else %} python2 {% endif %}/usr/local/bin/ycable
```
Comment (prgeor) [inline dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2:117]: can you please check if the niceness value is indeed what you are setting? you can use below command to check
```ps ax -o pid,ni,cmd```

**Review by prgeor — COMMENTED** [review]
@vdahiya12 can we rename the daemon to 'ycabled' just to have uniformity with other processes running in PMON? Also added other comments.

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 (hunk 7 above)
Comment (vdahiya12) [reply dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2:117]: 339903 -20 python3 /usr/local/bin/ycable

File: dockers/docker-platform-monitor/docker_init.j2 (hunk 4 above)
Comment (vdahiya12) [reply dockers/docker-platform-monitor/docker_init.j2:108]: admin@sonic:~$ sudo sonic-cfggen -h
Render configuration file from minigraph data and jinja2 template.

  -d, --from-db         read config from configdb
  Basically need to pass the argments to jinja file which can help is decide whether to initiate this daemon, according to 'DualToR' type

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 (hunk 3 above)
Comment (vdahiya12) [reply dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2:119]: I am thinking about it, all daemons have a priority from 1-10. Maybe need to have this as 7 ?

**Comment by vdahiya12** [comment]
Dependent on this https://github.com/Azure/sonic-platform-daemons/pull/219 and module update

File: rules/sonic-ycable.mk (hunk 5 above)
Comment (vdahiya12) [reply rules/sonic-ycable.mk]: fixed

File: rules/sonic-ycable.dep (hunk 6 above)
Comment (vdahiya12) [reply rules/sonic-ycable.dep]: fixed, removed PY2

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 [hunk 8]
```diff
@@ -108,6 +108,25 @@ dependent_startup=true
 dependent_startup_wait_for=rsyslogd:running
 {% endif %}
 
+{% if 'subtype' in DEVICE_METADATA['localhost'] and DEVICE_METADATA['localhost']['subtype'] == 'DualToR' %}
+{% if not skip_ycabled %}
+[program:ycabled]
+{% if delay_ycabled %}
+command=bash -c "sleep 30 && {% if API_VERSION == 3 and 'ycabled' not in python2_daemons %}nice -n -20 python3 {% else %} nice -n -20 python2 {% endif %}/usr/local/bin/ycabled"
```
Comment (prgeor) [inline dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2:115]: still py2 is used - 'python2_daemons'

**Review by prgeor — APPROVED** [review]

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 (hunk 8 above)
Comment (vdahiya12) [reply dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2]: fixed

**Review by prgeor — CHANGES_REQUESTED** [review]
i think you haven't tested the 'delay_ycabled' == true case?

File: dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2 [hunk 9]
```diff
@@ -110,10 +110,29 @@ dependent_startup=true
 dependent_startup_wait_for=rsyslogd:running
 {% endif %}
 
+{% if 'subtype' in DEVICE_METADATA['localhost'] and DEVICE_METADATA['localhost']['subtype'] == 'DualToR' %}
+{% if not skip_ycabled %}
+[program:ycabled]
+{% if delay_ycabled %}
+command=bash -c "sleep 30 nice -n -20 python3 /usr/local/bin/ycabled"
```
[the remark this answers is not in the document]
Comment (vdahiya12) [reply dockers/docker-platform-monitor/docker-pmon.supervisord.conf.j2]: Thank You for the catch, yes fixed now

**Review by prgeor — APPROVED** [review]
